### PR TITLE
Matrix save: Delimiter based on file extension

### PR DESCRIPTION
**Pull request by Lestropie**
Addresses #1184 if wanted.

Needs testing on locale with comma as decimal point separator, which I couldn't get working on my Windows system.

### CODE REVIEW

**Comment by thijsdhollander** [comment]
> Needs testing on locale with comma as decimal point separator, which I couldn't get working on my Windows system.

@maxpietsch @bjeurissen @dchristiaens ... maybe?

**Comment by Lestropie** [comment]
In retrospect, I don't think there's any need to check the locale of the output stream, as since #1323 we [force the C locale in `App::init()`](https://github.com/MRtrix3/mrtrix3/blob/3.0_RC3/core/app.cpp#L1156-L1157).

**Comment by Lestropie** [comment]
Converging Python matrix import code to behave equivalently to C++ wasn't quite as easy as I expected: `split()` doesn't accept multiple delimiters.

-----

@maxpietsch: The modified `save_transform()` & `load_transform()` that include a centre position are on reflection incongruent with #1603. See post-`dev`-merge code [here](https://github.com/MRtrix3/mrtrix3/blob/a4f6fffaedf314e2019a09ef7c2f5f2fd9672da1/core/math/math.h#L281-L305)

What I think would make more sense would be:

- For save:

    - Construct a `KeyValues` instance;

    - Insert into it an entry with key `centre` and value being the text string of the `Vector3`;

    - Pass that to the existing `save_transform()`.

- For load, have a wrapper function that:

    - Calls `load_transform()`;

    - Uses the existing `File::KeyValue::Reader` (which would require modification to support a comment symbol to strip from the start of each relevant line) to populate a `KeyValues` instance;

    - Validates the presence of the key "`centre`" & converts the value string to a `Vector3`;

    - Returns the transform and centre.

File: lib/mrtrix3/matrix.py [hunk 1]
```diff
@@ -106,7 +107,10 @@ def decode(line):
         line = regex_comments.split(line, maxsplit=1)[0]
       line = line.strip()
       if line:
-        data.append([dtype(a) for a in line.split(delimiter) if a])
+        if len(delimiter) == 1:
+          data.append([dtype(a) for a in line.split(delimiter) if a])
+        else:
+          data.append([dtype(a) for a in [''.join(g) for k, g in itertools.groupby(line, lambda c : c in delimiter) if not k ]])
```
Comment (jdtournier) [inline lib/mrtrix3/matrix.py]: 🤯

File: core/math/math.h [hunk 2]
```diff
@@ -19,13 +19,15 @@
 
 #include <cmath>
 #include <cstdlib>
+#include <locale>
```
Comment (jdtournier) [inline core/math/math.h]: Where is this required...?

File: lib/mrtrix3/matrix.py (hunk 1 above)
Comment (Lestropie) [reply lib/mrtrix3/matrix.py]: Rectangle indeed.

File: core/math/math.h (hunk 2 above)
Comment (Lestropie) [reply core/math/math.h]: Vestigial from previous attempts to detect whether or not it was safe to change the delimiter based on the decimal point symbol in use.

**Pull request merged**